### PR TITLE
Add Method decorators for improved Controller API

### PR DESCRIPTION
**Pull request by GDYendell**

### CODE REVIEW

File: malcolm/core/method.py [hunk 1]
```diff
@@ -106,3 +106,74 @@ def from_dict(cls, name, d):
...
+    Returns:
+        function: Updated function
+    """
+
+    def decorator(func):
+
+        if not hasattr(func, "Method"):
+            Method.wrap_method(func)
+
+        func.Method.takes = args
```
Comment (thomascobb) [inline malcolm/core/method.py]: should create map_meta = MapMeta("takes"), then for each pair of args a1, a2 call map_meta.add_element(a1, a2 is not OPTIONAL). If a2 is not in (REQUIRED, OPTIONAL), defaults[a1.name] = a2. Then call func.Method.set_function_takes(map_meta, defaults)

File: malcolm/core/method.py [hunk 2]
```diff
@@ -106,3 +106,74 @@ def from_dict(cls, name, d):
...
+        function: Updated function
+    """
+
+    def decorator(func):
+
+        if not hasattr(func, "Method"):
+            Method.wrap_method(func)
+
+        func.Method.returns = args
+
```
Comment (thomascobb) [inline malcolm/core/method.py]: Similar for returns but if a2 is not in (REQUIRED, OPTIONAL) then it is an error

**Comment by GDYendell** [comment]
The patch on TestMethod.test_decorators doesn't work for some reason, so it checks for a MapMeta instance. I want it to check that it is the mock I tried to set up but it seems to be ignoring them. Any idea why this is happening?

**Comment by GDYendell** [comment]
Coveralls implies the test isn't calling the decorator functions properly, which I guess is the problem. Not sure why though...

**Comment by thomascobb** [comment]
Can you rebase onto master please?

**Comment by GDYendell** [comment]
I need to remember to rebase to origin/master...

I think I know how to test it now so I will fix the coverage now.

**Comment by GDYendell** [comment]
OK, how is that?

**Pull request merged**